packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -354,49 +360,82 @@ export class Pipeline extends PipelineBase {
     ];
   }
 
-  private ensureReplicationBucketExistsFor(region?: string): s3.IBucket {
-    if (!region) {
-      return this.artifactBucket;
+  private ensureReplicationBucketExistsFor(action: IAction): CrossRegionInfo {
+    const pipelineStack = Stack.of(this);
+
+    let actionRegion: string | undefined;
+    let otherStack: Stack | undefined;
+
+    const actionResource = action.actionProperties.resource;
+    if (actionResource) {
+      const actionResourceStack = Stack.of(actionResource);
+      if (pipelineStack.region !== actionResourceStack.region) {
+        actionRegion = actionResourceStack.region;
+        otherStack = actionResourceStack;
+      }
+    } else {
+      actionRegion = action.actionProperties.region;
     }
 
+    // if actionRegion is undefined,
+    // it means the action is in the same region as the pipeline -
+    // so, just return the artifactBucket
+    if (!actionRegion) {
+      return {
+        artifactBucket: this.artifactBucket,
+      };
+    }
     // get the region the Pipeline itself is in
     const pipelineRegion = this.requireRegion();
-
-    // if we already have an ArtifactStore generated for this region, or it's the Pipeline's region, nothing to do
-    if (region === pipelineRegion) {
-      return this.artifactBucket;
+    // if the action is in the same region as the pipeline, nothing to do
+    if (actionRegion === pipelineRegion) {
+      return {
+        artifactBucket: this.artifactBucket,
+      };
     }
 
-    let replicationBucket = this.crossRegionReplicationBuckets[region];
-    if (!replicationBucket) {
-      const pipelineStack = Stack.of(this);
-      const pipelineAccount = pipelineStack.account;
-      if (Token.isUnresolved(pipelineAccount)) {
-        throw new Error("You need to specify an explicit account when using CodePipeline's cross-region support");
-      }
+    let crossRegionSupport = this._crossRegionSupport[actionRegion];
+    if (!crossRegionSupport) {
+      // if we have a stack from the resource passed - use that!
+      if (otherStack) {
+        const replicationBucket = new s3.Bucket(otherStack, 'CrossRegionCodePipelineReplicationBucket', {
```

Comment:
Should probably reuse if there is already a replication bucket in this stack. Also "salt" the ID with some portion of a UUID or random string to ensure user constructs do not conflict with it.

Reply:
Done.